### PR TITLE
Refactor Range header parsing into a separate function and fix handling of zero-sized files

### PR DESCRIPTION
**Pull request by s-ludwig**
Changes the internal representation of a range to be end-exclusive (consistent with D slices) to avoid unsigned integer wrap-around for zero sized files.

See https://github.com/vibe-d/vibe.d/pull/2481#discussion_r504012709.

### CODE REVIEW

File: http/vibe/http/fileserver.d [hunk 1]
```diff
@@ -709,3 +678,55 @@ unittest
...
+		throw new HTTPStatusException(HTTPStatus.rangeNotSatisfiable);
+	}
+
+	return ret;
+}
+
+unittest {
+	auto res = createTestHTTPServerResponse();
+	assertThrown(parseRangeHeader("bytes=2-1", 10, res));
+	assertThrown(parseRangeHeader("bytes=10-10", 10, res));
```
Comment (WebFreak001) [inline http/vibe/http/fileserver.d:722]: I believe this is a valid range (10-10 would simply mean, just the one byte at index `10`) and thus should not throw.

See 2.1 in the RFC which only says:

```
   A byte-range-spec is invalid if the last-byte-pos value is present
   and less than the first-byte-pos.
```

plus it's actually used as an example here:

```
   o  The first and last bytes only (bytes 0 and 9999):
        bytes=0-0,-1
```

**Review by WebFreak001 — CHANGES_REQUESTED** [review]

File: http/vibe/http/fileserver.d (hunk 1 above)
Comment (s-ludwig) [reply http/vibe/http/fileserver.d:722]: Wait, but index 10 is outside of 10 bytes isn't it?

Comment (s-ludwig) [reply http/vibe/http/fileserver.d:722]: Or do you mean that it should get moved into the range as `9-9`?

Comment (WebFreak001) [reply http/vibe/http/fileserver.d:722]: oh sorry you are right about that, I wasn't looking at the size there

**Pull request merged**